lighthouse-core/lib/tracehouse/trace-processor.js

```diff
@@ -66,10 +66,68 @@ class TraceProcessor {
   }
 
   /**
+   * This method sorts a group of trace events that have the same timestamp. We want to...
+   *
+   * 1. Put E events first, we finish off our existing events before we start new ones.
+   * 2. Order B/X events by their duration, we want parents to start before child events.
+   * 3. If we don't have any of this to go on, just use the position in the original array (stable sort).
+   *
+   * @param {number[]} tsGroupIndices
+   * @param {number[]} timestampSortedIndices
+   * @param {number} indexOfTsGroupIndicesStart
+   * @param {LH.TraceEvent[]} traceEvents
+   * @return {number[]}
+   */
+  static _sortTimestampEventGroup(
+    tsGroupIndices,
+    timestampSortedIndices,
+    indexOfTsGroupIndicesStart,
+    traceEvents
+  ) {
+    const eEventIndices = tsGroupIndices.filter(index => traceEvents[index].ph === 'E');
+    const bxEventIndices = tsGroupIndices.filter(index => traceEvents[index].ph === 'X' ||
+      traceEvents[index].ph === 'B');
+    const otherEventIndices = tsGroupIndices.filter(index => !eEventIndices.includes(index) &&
+      !bxEventIndices.includes(index));
+
+    const effectiveDuration = new Map();
+    for (const index of bxEventIndices) {
+      const event = traceEvents[index];
+      if (event.ph === 'X') {
+        effectiveDuration.set(index, event.dur);
+      } else {
+        // Find the first 'E' event that matches our name.
+        const startIndex = indexOfTsGroupIndicesStart + tsGroupIndices.length;
```

Comment:
maybe add a mention that this is starting after the current group of ts events, it takes a minute or two to see that :)